A41-xds-rbac.md

```diff
@@ -62,107 +62,134 @@ Supporting the RBAC HTTP filter on server-side leverages [A36: xDS-Enabled
+incorrect value, the Listener must be NACKed. For simplicity, this behavior
+applies independent of the Listener type (both client-side and server-side).
+
 The core policy matching logic should be split into an "RBAC engine" to allow
-reuse with non-xDS environments. However, there is no requirement that the RBAC
-engine have a specialized API; it could simply be an interceptor.
+internal reuse with a non-xDS API. Any non-xDS API will not be RBAC, so this
+reuse is merely an implementation detail. The API is not part of this gRFC.
+There is no requirement that the RBAC engine have a specialized API; it could
+simply be an interceptor.
 
 The [RBAC policy][] has many fields. All current (Envoy-implemented) fields will
 be considered in gRPC. However, some fields may have pre-determined values or
 behavior. At this time, if the `RBAC.action` is `Action.LOG` then the policy
 will be completely ignored, as if RBAC was not configurated. CEL fields are not
-supported, so if `Policy.condition` or `Policy.checked_condition` is present the
-XdsClient must NACK the configuration.
+supported, so `Policy.condition` and `Policy.checked_condition` must cause a
+validation failure if present. It is also a validation failure if Permission or
+Principal has a `header` matcher for a `grpc-`-prefixed header name or
+`:scheme`. As described in A39, validation failures for filter configuration
+causes the listener to be NACKed.
 
 The following fields of `Permission` may not be obvious how they map to gRPC:
 
 | Permission Field | gRPC Equivalent |
 | ---------------- | --------------- |
-| header           | Metadata (with caveats) |
-| url_path         | Fully-qualified RPC method name with leading slash |
+| header           | Metadata (with caveats below) |
+| url_path         | Fully-qualified RPC method name with leading slash. Same as `:path` header |
 | destination_ip   | Local address for this connection |
 | destination_port | Local port for this connection |
 | metadata         | Hard-coded as empty; never matches |
 | requested_server_name | Hard-coded as empty string |
 
-Be strongly aware that Envoy Metadata has no relation to gRPC's Metadata.
 Because matching supports NOT, the matcher must still be processed even if a
 rule contains references to things that don't generally match; it is not trivial
 to "optimize out" the never-matching rules.
 
-The `header` field is not entirely 1:1 with gRPC Metadata. To begin with, gRPC
-Metadata isn't 100% consistent cross-language in its handling of hop-to-hop
-headers (e.g., `TE`) and pseudo-headers. For this design, `headers` can include
-`:method`, `:scheme`, `:authority`, and`:path` matchers and they must match the
-values received on-the-wire independent of whether they are stored in Metadata
-or in separate APIs. It is unspecified whether hop-to-hop headers are matched.
+The `header` field is not entirely 1:1 with gRPC Metadata, in part because which
```

Comment:
Some of the behavior described here is going to be a bit of a challenge for C-core, since the xDS HTTP filters will sit higher in the stack than the http_server filter, which removes a lot of the fields that we want to see here.  In particular, it removes `:method` and `Content-type`, so they will no longer be visible by the time we get to the RBAC filter.  CC @yashykt 

Can you remind me how Java and Go handle these headers?  Do they simply get returned to the application?  CC @dfawley

Reply:
`:method` will be hard-coded to `POST` in Java. `Content-Type` happens to still be present in Metadata in Java. Netty does validate it on a low-level, but doesn't remove it because 1) it doesn't remove any headers and 2) the higher levels would need it if we added support for the `+proto`/`+json` part.

Reply:
In Go, we currently strip `:method` (we validate it is `POST`), but `content-type` is passed through to the application.

Reply:
I think we've come up with a reasonable way to handle `:method` in C-core, but I'm not yet sure how we're going to handle `content-type`.  We need to think about this some more.